### PR TITLE
Finalize the callback API

### PR DESCRIPTION
**Pull request by martintmk**
### Details on the issue fix or feature implementation

This PR finalizes and cleans-up the callback API.

- Introducing `NoOutcomeGenerator` and dropping `TimeoutGenerator`
- Introducing `TResult` based class for Generators and Events
- To verify the API a new generic `ResilienceStrategyOptions<TResult>` are introduced. The user will use these when they care only about specific result type. These can be easily converted to `ResilienceStrategyOptions`. 

I hope with this PR the callback API is almost done only minor changes (if any) will be done later.

### Confirm the following

- [x]  I started this PR by branching from the head of the default branch
- [x]  I have targeted the PR to merge into the default branch
- [x]  I have included unit tests for the issue/feature
- [x]  I have successfully run a local build

### CODE REVIEW

File: src/Polly.Core/Strategy/OutcomeEvent.cs [hunk 1]
```diff
@@ -78,16 +66,39 @@ public OutcomeEvent<TArgs> Register<TResult>(Func<Outcome<TResult>, TArgs, Value
...
+    /// Adds a callback for the specified result type.
+    /// </summary>
+    /// <typeparam name="TResult">The result type to add a callbacks for.</typeparam>
+    /// <param name="configure">Action that configures the callbacks.</param>
+    /// <returns>The current updated instance.</returns>
+    public OutcomeEvent<TArgs> ConfigureCallbacks<TResult>(Action<OutcomeEvent<TArgs, TResult>> configure)
+    {
+        Guard.NotNull(configure);
+
+        if (!_callbacks.ContainsKey(typeof(TResult)))
```
Comment (martincostello) [inline src/Polly.Core/Strategy/OutcomeEvent.cs:82]: Use `!TryGetValue()` here so we don't need to index on line 87?

File: src/Polly.Core/Strategy/OutcomeGenerator.TResult.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+/// <typeparam name="TResult">The result type that this event handles.</typeparam>
+public sealed class OutcomeGenerator<TArgs, TValue, TResult>
+    where TArgs : IResilienceArguments
+{
+    private Func<Outcome<TResult>, TArgs, ValueTask<TValue>>? _generator;
+
+    /// <summary>
+    /// Gets a value indicating whether the generator is empty.
+    /// </summary>
+    public bool IsEmpty => _generator == null;
```
Comment (martincostello) [inline src/Polly.Core/Strategy/OutcomeGenerator.TResult.cs:23]:
```suggestion
    public bool IsEmpty => _generator is null;
```

File: src/Polly.Core/Strategy/OutcomeGenerator.cs [hunk 3]
```diff
@@ -1,17 +1,18 @@
 using System;
+using System.Collections.Generic;
 using Polly.Strategy;
 
 namespace Polly.Strategy;
 
 /// <summary>
-/// The base class for all generators that generate a value based on the <see cref="Outcome{TResult}"/>.
+/// Generators that generates a value based on the <see cref="Outcome{TResult}"/>.
```
Comment (martincostello) [inline src/Polly.Core/Strategy/OutcomeGenerator.cs:8]:
```suggestion
/// A class that generates a value based on the <see cref="Outcome{TResult}"/>.
```

File: src/Polly.Core/Strategy/OutcomeGenerator.cs [hunk 4]
```diff
@@ -22,53 +23,83 @@ public sealed partial class OutcomeGenerator<TArgs, TValue>
...
     public OutcomeGenerator<TArgs, TValue> SetGenerator(Func<Outcome, TArgs, ValueTask<TValue>> generator)
     {
         Guard.NotNull(generator);
 
-        _generators[typeof(AnyResult)] = generator;
+        return ConfigureGenerator<object>(g => g.SetGenerator((outcome, args) => generator(outcome.AsOutcome(), args)));
+    }
+
+    /// <summary>
+    /// Adds a result generator for specific result type.
```
Comment (martincostello) [inline src/Polly.Core/Strategy/OutcomeGenerator.cs:73]:
```suggestion
    /// Adds a result generator for a specific result type.
```

File: src/Polly.Core/Strategy/OutcomeGenerator.cs [hunk 5]
```diff
@@ -22,53 +23,83 @@ public sealed partial class OutcomeGenerator<TArgs, TValue>
...
+    /// Adds a result generator for all result types including the void-based results.
+    /// </summary>
+    /// <typeparam name="TResult">The result type to add a generator for.</typeparam>
+    /// <param name="configure">The callbacks that configures the generator.</param>
+    /// <returns>The current updated instance.</returns>
+    public OutcomeGenerator<TArgs, TValue> ConfigureGenerator<TResult>(Action<OutcomeGenerator<TArgs, TValue, TResult>> configure)
+    {
+        Guard.NotNull(configure);
+
+        if (!_generators.ContainsKey(typeof(TResult)))
```
Comment (martincostello) [inline src/Polly.Core/Strategy/OutcomeGenerator.cs:97]: `!TryGetValue()` too?

File: src/Polly.Core/Strategy/OutcomeGenerator.cs [hunk 6]
```diff
@@ -80,17 +111,16 @@ public sealed partial class OutcomeGenerator<TArgs, TValue>
     /// <returns>Handler instance or null if no generators are registered.</returns>
     public Handler? CreateHandler(TValue defaultValue, Predicate<TValue> valueValidator)
     {
-        var pairs = _generators.ToArray();
+        var pairs = _generators
+            .Select(pair => new KeyValuePair<Type, object?>(pair.Key, pair.Value.handlerFactory()))
+            .Where(pair => pair.Value != null)
```
Comment (martincostello) [inline src/Polly.Core/Strategy/OutcomeGenerator.cs:116]:
```suggestion
            .Where(pair => pair.Value is not null)
```

File: src/Polly.Core/Strategy/OutcomePredicate.cs [hunk 7]
```diff
@@ -161,19 +161,25 @@ public OutcomePredicate<TArgs> ConfigurePredicates<TResult>(Action<OutcomePredic
...
-        configure(predicate);
+        configure((OutcomePredicate<TArgs, TResult>)_predicates[typeof(TResult)].predicate);
+        return this;
+    }
+
+    /// <summary>
+    /// Sets the predicates for the specified result type.
+    /// </summary>
+    /// <typeparam name="TResult">The result type to add a predicate for.</typeparam>
+    /// <param name="predicate">The configured predicates.</param>
```
Comment (martincostello) [inline src/Polly.Core/Strategy/OutcomePredicate.cs:177]: Mix of plural (_"predicates"_) and singular (_"a predicate"_) here.

Comment (martintmk) [reply src/Polly.Core/Strategy/OutcomePredicate.cs]: Renamed.

File: src/Polly.Core/Strategy/OutcomeGenerator.cs (hunk 6 above)
Comment (martintmk) [reply src/Polly.Core/Strategy/OutcomeGenerator.cs]: Done.

File: src/Polly.Core/Strategy/OutcomeGenerator.cs (hunk 5 above)
Comment (martintmk) [reply src/Polly.Core/Strategy/OutcomeGenerator.cs]: Yup, it's better.

File: src/Polly.Core/Strategy/OutcomeGenerator.cs (hunk 4 above)
Comment (martintmk) [reply src/Polly.Core/Strategy/OutcomeGenerator.cs:73]: Done.

File: src/Polly.Core/Strategy/OutcomeGenerator.cs (hunk 3 above)
Comment (martintmk) [reply src/Polly.Core/Strategy/OutcomeGenerator.cs]: Done.

File: src/Polly.Core/Strategy/OutcomeGenerator.TResult.cs (hunk 2 above)
Comment (martintmk) [reply src/Polly.Core/Strategy/OutcomeGenerator.TResult.cs]: Done.

File: src/Polly.Core/Strategy/OutcomeEvent.cs (hunk 1 above)
Comment (martintmk) [reply src/Polly.Core/Strategy/OutcomeEvent.cs]: Done.

**Review by martincostello — APPROVED** [review]

**Pull request merged**